src/libs/safe-path/src/scoped_dir_builder.rs

```diff
@@ -0,0 +1,215 @@
+    pub fn mode(&mut self, mode: u32) -> &mut Self {
+        self.mode = mode & DIRECTORY_MODE_MASK;
+        self
+    }
+
+    /// Creates the specified directory with the options configured in this builder.
+    ///
+    /// The `path` must be a subdirectory of `ScopedDirBuilder::root`, otherwise error will be
+    /// returned. It is considered an error if the directory already exists unless recursive mode
+    /// is enabled.
+    pub fn create<P: AsRef<Path>>(&self, path: P) -> Result<PinnedPathBuf> {
+        let path = scoped_join("/", path)?;
+        self.do_mkdir(&path)
+    }
+
+    /// Creates sub-directory with the options configured in this builder.
+    ///
+    /// It is considered an error if the directory already exists unless recursive mode is enabled.
+    pub fn create_sub_directory<P: AsRef<Path>>(&self, path: P) -> Result<PinnedPathBuf> {
+        let path = scoped_join(&self.root, path)?;
+        self.do_mkdir(&path)
+    }
+
+    fn do_mkdir(&self, path: &PathBuf) -> Result<PinnedPathBuf> {
+        let mut suffix = path.strip_prefix(&self.root).map_err(|_| {
+            Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", path.display()),
+            )
+        })?;
+
+        let mut root = self.root.clone();
+        if suffix.file_name().is_none() {
+            if !self.recursive {
+                return Err(Error::new(
+                    ErrorKind::AlreadyExists,
+                    "directory already exists",
+                ));
+            } else {
+                return PinnedPathBuf::from_path(root);
+            }
+        }
+
+        if !self.recursive {
+            if let Some(parent) = path.parent() {
+                root = root.join(parent);
+            }
+            // Safe to unwrap() because we have verified `suffix` is not empty.
+            suffix = Path::new(suffix.file_name().unwrap());
+        }
+
+        for comp in suffix {
+            let dir = PinnedPathBuf::from_path(&root)?;
+            if !dir.is_dir() {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!("Expected directory at {}", root.display()),
+                ));
+            } else if dir.target() != &root {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!(
+                        "Directory changed from {} to {}, possible attack",
+                        root.display(),
+                        dir.target().display()
+                    ),
+                ));
+            }
+
+            let name = CString::new(comp.as_bytes()).map_err(|_e| {
+                Error::new(
+                    ErrorKind::Other,
+                    format!("Invalid path: {}", root.display()),
+                )
+            })?;
+            let res =
+                unsafe { libc::mkdirat(dir.path_handle().raw_fd(), name.as_ptr(), self.mode) };
```

Comment:
Probably nicer to make `mkdirat` a method of `PathHandle`, handling the `OsStr` to `CString` conversion in there.

Reply:
fixed